substrate-archive/src/database/queries.rs

```diff
@@ -178,22 +152,137 @@ pub(crate) async fn get_versions(conn: &mut PgConnection) -> Result<Vec<u32>> {
 		.collect())
 }
 
-/// Get all the blocks queued for execution in the background task queue.
-pub(crate) async fn get_all_blocks<B: BlockT + DeserializeOwned>(
-	conn: &mut PgConnection,
-) -> Result<impl Iterator<Item = Result<B>>> {
-	#[allow(clippy::toplevel_ref_arg)]
-	let blocks = sqlx::query_as!(Bytes, "SELECT data FROM _background_tasks WHERE job_type = 'execute_block'",)
+pub(crate) async fn missing_storage_blocks(conn: &mut sqlx::PgConnection) -> Result<Vec<u32>> {
+	let blocks: Vec<u32> = sqlx::query_as!(BlockNum,
+	   r#"SELECT block_num AS "block_num!" FROM
+            (SELECT block_num FROM blocks EXCEPT
+                SELECT (HEX_TO_INT(LTRIM(data->'block'->'header'->>'number', '0x'))) AS block_num FROM _background_tasks WHERE job_type = 'execute_block') AS maybe_missing
+        WHERE NOT EXISTS
+	        (SELECT block_num FROM storage WHERE storage.block_num = maybe_missing.block_num)
+		ORDER BY block_num"#
+	)
 		.fetch_all(conn)
-		.await?;
+		.await?
+		.into_iter()
+		.map(|r| r.block_num as u32)
+		.collect();
+	Ok(blocks)
+}
+
+/// Get full blocks in pages
+pub(crate) fn blocks_paginated<'a>(
+	conn: &'a mut sqlx::PgConnection,
+	nums: &'a [u32],
+	limit: usize,
+) -> impl Stream<Item = Result<Vec<BlockModel>>> + 'a {
+	let nums: Vec<i32> = nums.iter().filter_map(|n| i32::try_from(*n).ok()).collect();
+
+	Box::pin(try_stream! {
+		for page in nums.chunks(limit) {
+			let mut query = String::from("SELECT * FROM blocks WHERE block_num IN (");
+			for (i, num) in page.iter().enumerate() {
+				itoa::fmt(&mut query, *num)?;
```

Comment:
are you using itoa for performance reasons here instead of `format!()`?

Reply:
Yes! This needs to format alotof block numbers